src/SAML2/WebSSO/XML.hs

```diff
@@ -861,17 +873,18 @@ parseIdPMetadataHead el@(Element tag attrs _) = do
     issueruri :: ST <- maybe (throwError "no issuer") pure (Map.lookup "entityID" attrs)
     Issuer <$> parseURI' issueruri
   _edRequestURI :: URI <- do
-    let cur = fromNode $ NodeElement el
-        target :: [ST]
-        target =
-          cur $// element "{urn:oasis:names:tc:SAML:2.0:metadata}IDPSSODescriptor"
-            &/ element "{urn:oasis:names:tc:SAML:2.0:metadata}SingleSignOnService"
-            >=> attributeIs "Binding" "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
-            >=> attribute "Location"
-    case parseURI' <$> target of
-      [Right uri] -> pure uri
-      [Left msg] -> throwError $ "bad request uri: " <> msg
-      bad -> throwError $ "unexpected request uri: " <> show (target, bad)
+    target :: ST <-
+      [fromNode (NodeElement el)]
+        & ( findSome "IDPSSODescriptor element" (descendant >=> element "{urn:oasis:names:tc:SAML:2.0:metadata}IDPSSODescriptor")
+              >=> findSome "SingleSignOnService" (child >=> element "{urn:oasis:names:tc:SAML:2.0:metadata}SingleSignOnService")
+              >=> findSome "\"Binding\" attribute with value \"HTTP-POST\"" (attributeIs "Binding" "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST")
+              >=> getSingleton "\"Binding\" attribute with value \"HTTP-POST\""
```

Comment:
you can skip the `findSome` if you then call `getSingleton`, no?

```suggestion
              >=> attributeIs "Binding" "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
              >=> getSingleton "\"Binding\" attribute with value \"HTTP-POST\""
```

Reply:
I updated the error message, but I couldn't find a way to skip `findSome` or refactor it better